metaflow/plugins/tag_cli.py

```diff
@@ -0,0 +1,531 @@
+    try:
+        namespace(user_namespace)
+        Flow(pathspec=flow_name)
+    except MetaflowNotFound:
+        raise CommandException(
+            "No execution found for *%s*. Please run the flow before tagging."
+            % flow_name
+        )
+
+    except MetaflowNamespaceMismatch:
+        raise CommandException(
+            "No execution found for *%s* in namespace *%s*."
+            % (flow_name, user_namespace)
+        )
+
+    # throw an error with message to include latest run-id when run_id is None
+    if run_id is None:
+        latest_run_id = Flow(pathspec=flow_name).latest_run.id
+        msg = (
+            "Please specify a run-id using --run-id.\n"
+            "*%s*'s latest execution in namespace *%s* has id *%s*."
+            % (flow_name, user_namespace, latest_run_id)
+        )
+        raise CommandException(msg)
+    run_id_parts = run_id.split("/")
+    if len(run_id_parts) == 1:
+        run_name = "%s/%s" % (flow_name, run_id)
+    else:
+        raise CommandException("Run-id *%s* is not a valid run-id" % run_id)
+
+    # handle error messaging for three cases
+    # 1. our user makes a typo in --run-id
+    # 2. our user tries to mutate tags on a meson/production run
+    # 3. our user's --run-id does not exist in the default/specified namespace
+    try:
+        namespace(user_namespace)
+        run = Run(pathspec=run_name)
+    except MetaflowNotFound:
+        raise CommandException("No run *%s* found" % run_name)
+    except MetaflowNamespaceMismatch:
+        namespace(None)
+        run = Run(pathspec=run_name)
+
+        if any([_is_prod_token(tag) for tag in run.system_tags]):
+            raise CommandException(
+                "Modifying a scheduled run's tags is currently not allowed."
+            )
+
+        msg = "Run *%s* does not belong to namespace *%s*\n" % (
+            run_name,
+            user_namespace,
+        )
+
+        user_identifier_tags = [
+            tag for tag in run.system_tags if tag.startswith("user:")
+        ]
+        if user_identifier_tags:
+            msg += (
+                "You can choose any of the following tags associated with *%s* "
+                "to use with option *--namespace*\n" % run_name
+            )
+            msg += "\n".join(["\t*%s*" % tag for tag in user_identifier_tags])
+
+        raise CommandException(msg)
+
+    return run
+
+
+@click.group()
+def cli():
+    pass
+
+
+@cli.group(help="Commands related to Metaflow tagging.")
+def tag():
+    pass
+
+
+@tag.command("add", help="Add tags to a run.")
+@click.option(
+    "--run-id",
+    required=False,  # set False here so we can throw a better error message
+    default=None,
+    type=str,
+    help="Run ID of the specific run to tag. [required]",
+)
+@click.option(
+    "--namespace",
+    "user_namespace",
+    required=False,
+    default=None,
+    type=str,
+    help="Change namespace from the default (your username) to the one specified.",
```

Comment:
@romain-intel @tuulos should we support the global namespace?

Reply:
You mean like `namespace = None` == `all namespaces` ? I don't mind but it seems a little possibly dangerous.

Reply:
more like `--global-namespace`. We don't have to do it now - if a use case presents itself, we can always add it later.

Reply:
OK resolving this convo.  We know where to look later.